### PR TITLE
Otchet/dr 2711 patch dataset description in UI

### PR DESCRIPTION
**Pull request by otchet-broad**
Adds dataset description editing to the UI.

(a new PR seemed to do better...)

### CODE REVIEW

**Comment by otchet-broad** [comment]
FYSA: This PR code is tailored to the implementation for datasets only.  For example, canEdit is computed inside of the DescriptionView component.  In DR-2712, this is refactored so the calculation of canEdit occurs on the resource using the description component.

File: src/components/DescriptionView.jsx [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+import { Button, Box, Grid, IconButton, TextField, Typography } from '@mui/material';
+import Edit from '@mui/icons-material/Edit';
+import { showNotification } from 'modules/notifications';
+import UndoIcon from '@mui/icons-material/Undo';
+import SaveIcon from '@mui/icons-material/Save';
+import { withStyles } from '@mui/styles';
+import { patchDatasetDescription } from 'actions';
+import { DatasetRoles } from '../constants';
+
+const styles = (theme) => ({
```
Comment (nmalfroy) [inline src/components/DescriptionView.jsx]: given that this is a new component, could you make this a `tsx`?

File: src/components/DescriptionView.jsx [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+});
+
+const maxLength = 2047;
+
+const descriptionTooLongError = {
+  response: {
+    status: 'Description too long',
+    data: {
+      message: '',
+      errorDetail: 'Exceeds 2047 characters.  Please revise and save again.',
```
Comment (nmalfroy) [inline src/components/DescriptionView.jsx]: Might want to use the variable in here
```suggestion
      errorDetail: `Exceeds ${maxLength} characters.  Please revise and save again.`,
```

File: src/components/DescriptionView.jsx [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+    color: theme.palette.grey[500],
+    margin: theme.spacing(1),
+    '&:hover': {
+      backgroundColor: theme.palette.primary.light,
+      boxShadow: 'none',
+    },
+  },
+});
+
+const maxLength = 2047;
```
Comment (nmalfroy) [inline src/components/DescriptionView.jsx]: we typically snake case constants
```suggestion
const MAX_LENGTH = 2047;
```

File: src/components/DescriptionView.jsx [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+      message: '',
+      errorDetail: 'Exceeds 2047 characters.  Please revise and save again.',
+    },
+  },
+};
+class DescriptionView extends React.PureComponent {
+  constructor(props) {
+    super(props);
+    const descriptionInitialValue = props.description;
+    this.state = {
```
Comment (nmalfroy) [inline src/components/DescriptionView.jsx]: We've been moving towards using React effects for components. More of an FYI. We can always convert it over afterwards but just as a heads up

**Review by nmalfroy — CHANGES_REQUESTED** [review]
Looks great!  My main thing is that the new component should use typescript

File: src/components/DescriptionView.jsx (hunk 1 above)
Comment (otchet-broad) [reply src/components/DescriptionView.jsx]: Refactoring now.

File: src/components/DescriptionView.tsx [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+type DescriptionViewProps = {
+  classes: ClassNameMap;
+  description: string;
+  updateDescriptionFn: any;
+  userRoles: Array<string>;
+};
+
+type DescriptionViewState = {
+  hasDescriptionChanged: boolean;
+  descriptionValue: string | null;
```
Comment (otchet-broad) [inline src/components/DescriptionView.tsx:79]: FYSA: nulls are refactored to undefined in DR-2712

File: src/components/DescriptionView.tsx [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+      this.onExitEdit();
+    }
+  }
+
+  textFieldRef: React.RefObject<any>;
+
+  descriptionChanged(newDescription: string | null, originalDescription: string | null) {
+    if (newDescription && newDescription.length > MAX_LENGTH) {
+      showNotification(descriptionTooLongError);
+      this.setState({ descriptionValue: newDescription.substring(0, 2046) });
```
Comment (nmalfroy) [inline src/components/DescriptionView.tsx]: should this be:
```suggestion
      this.setState({ descriptionValue: newDescription.substring(0, MAX_LENGTH - 1) });
```

File: src/components/DescriptionView.tsx [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+    updateDescriptionFn(descriptionText);
+  }
+
+  render() {
+    const { classes, description, userRoles } = this.props;
+    const { hasDescriptionChanged, descriptionValue, isEditing } = this.state;
+    const canEdit = userRoles.includes(DatasetRoles.STEWARD);
+
+    return (
+      <>
```
Comment (nmalfroy) [inline src/components/DescriptionView.tsx:165]: oh interesting: I've never seen this done with the empty bracket rather than arrays.  I like this

**Review by nmalfroy — APPROVED** [review]
This looks great!  Thanks for the refactor

File: src/components/DescriptionView.tsx (hunk 6 above)
Comment (otchet-broad) [reply src/components/DescriptionView.tsx]: Yes!  Good catch.  Thank you!

File: src/components/DescriptionView.test.tsx [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+          updateDescriptionFn={(text: any) => {
+            initialState.dataset.description = text;
+          }}
+        />
+      </ThemeProvider>
+    </Router>,
+  );
+});
+
+describe('DescriptionView', () => {
```
Comment (okotsopoulos) [inline src/components/DescriptionView.test.tsx:38]: I have little-to-no practical experience in this repo / UI technologies, I appreciate how easy to follow your tests are!

**Review by okotsopoulos — APPROVED** [review]
Great work on this!  I really appreciate how you took a ticket that was meant to be a simple entrypoint into our UI code and turned it into an opportunity to make thoughtful, tangible improvements.

A few meta-comments / preferences.

- Reviewing with the destination branch's commits interwoven through the diff view is a little confusing.  I *think* we tend to prefer merging rather than rebasing when bringing a feature branch up-to-date (@nmalfroy, please keep me honest if that's inaccurate).

- I find reviewing PRs easier when they contain fleshed-out descriptions, even if this context is redundant with what's in the ticket or what's been demoed in person.  [This article](https://mtlynch.io/code-review-love/) is a favorite of the team's, with technique 2 being relevant:

>Your changelist description should summarize any background knowledge the reader needs. You might have a code reviewer in mind when you write the description, but they don’t necessarily have the context you imagine. Besides, your other teammates might need to read this changelist as well, and readers in the future should understand your intentions when they look back on the change history.

**Pull request merged**